src/efftester.ml

```diff
@@ -652,107 +735,149 @@ and appRules env t eff size =
-	    let goal_type = subst sub' goal_type in
-	    let argTypes = try getArgTypes goal_type (subst sub t)
-                           with Failure exc ->
-                             print_endline ("s is " ^ (typetostr ~effannot:true s));
-                             print_endline ("sub is " ^ (Print.list (Print.pair
-                                                                       (fun id -> "'a" ^ (string_of_int id))
-                                                                       (typetostr ~effannot:true)) sub));
-                             print_endline ("(subst sub s) is " ^ (typetostr ~effannot:true (subst sub s)));
-                             print_endline ("t is " ^ (typetostr ~effannot:true t));
-                             failwith exc in
-	    let first_eff_index = first_eff goal_type in
- 	    Gen.((if first_eff_index = 0
-	          then return 0
-      	          else int_bound (first_eff_index-1)) >>= fun n ->
-   		    apply fTerm goal_type n no_eff argTypes))
+  let mgu s t =
+    let rec loop = function
+      | [] -> None
+      | r :: rs ->
+        (match unify r t with
+        | Sol u -> Some u
+        | No_sol -> loop rs)
     in
-
-    let normalized_t = normalize_eff t in
-    let suitableVariables = lookupReturn (normalized_t) env in (* this returns a set of cand. sans effects *)
-    let f_type_map =
-      let rec acc_rt_and_effect eff ty = (* check that all effects are leq eff and that rt is compat *)
-	(if types_compat (normalize_eff ty) normalized_t (* (sans effects) types agree here *)
-	 then types_compat ty t
-	 else true)
-	&& match ty with
-	  | Fun (_,e,ty') -> eff_leq e eff && acc_rt_and_effect eff ty'
-	  | _           -> true
+    loop (returnTypes s)
+  in
+  let rec getArgTypes s t =
+    match s with
+    | s when types_compat s t -> []
+    | Fun (s', _, t') -> s' :: getArgTypes t' t
+    | s ->
+      failwith
+        ("getArgTypes: should not happen  s is "
+        ^ typetostr ~effannot:true s
+        ^ " t is "
+        ^ typetostr ~effannot:true t)
+  in
+  (* returns the index of the first effect - or else the number of arguments *)
+  let rec first_eff = function
+    | s when types_compat s t || types_compat t s -> 0
+    | Fun (_, e, t) -> if e = no_eff then 1 + first_eff t else 1
+    | s -> failwith ("first_eff: should not happen  " ^ typetostr ~effannot:true s)
+  in
+  (* recursively build application term argument by argument *)
+  let rec apply term rType n effacc = function
+    | [] -> Gen.return (Some term)
+    | arg :: args ->
+      let myeff = if n = 0 then eff else no_eff in
+      (* arg 'n' may have effect 'eff' *)
+      Gen.( >>= )
+        (listPermuteTermGenOuter env arg myeff (size / 2))
+        (function
+          | None -> Gen.return None
+          | Some a ->
+            (match rType with
+            | Fun (argType, funeff, newRType) ->
+              let my_actual_eff = eff_join funeff (imm_eff a) in
+              (* actual effect *)
```

Comment:
The comment on this line is not terribly readable. Maybe you could try to move it before the `in` to force it on the previous line? (I have seen a few similar cases.)